iocore/net/quic/QUICUnbidirectionalStream.cc

```diff
@@ -0,0 +1,760 @@
+    ink_assert(len != 0);
+
+    if (this->_write_vio.nbytes == static_cast<int64_t>(this->_send_offset + len)) {
+      fin = true;
+    }
+  }
+
+  Ptr<IOBufferBlock> block = make_ptr<IOBufferBlock>(reader->get_current_block()->clone());
+  block->consume(reader->start_offset);
+  block->_end = std::min(block->start() + len, block->_buf_end);
+  ink_assert(static_cast<uint64_t>(block->read_avail()) == len);
+
+  // STREAM - Pure FIN or data length is lager than 0
+  // FIXME has_length_flag and has_offset_flag should be configurable
+  frame = QUICFrameFactory::create_stream_frame(buf, block, this->_id, this->_send_offset, fin, true, true, this->_issue_frame_id(),
+                                                this);
+  if (!this->_state.is_allowed_to_send(*frame)) {
+    QUICStreamDebug("Canceled sending %s frame due to the stream state", QUICDebugNames::frame_type(frame->type()));
+    return frame;
+  }
+
+  if (!pure_fin) {
+    int ret = this->_remote_flow_controller.update(this->_send_offset + len);
+    // We cannot cancel sending the frame after updating the flow controller
+
+    // Calling update always success, because len is always less than stream_credit
+    ink_assert(ret == 0);
+
+    QUICStreamFCDebug("[REMOTE] %" PRIu64 "/%" PRIu64, this->_remote_flow_controller.current_offset(),
+                      this->_remote_flow_controller.current_limit());
+    if (this->_remote_flow_controller.current_offset() == this->_remote_flow_controller.current_limit()) {
+      QUICStreamDebug("Flow Controller will block sending a STREAM frame");
+    }
+
+    reader->consume(len);
+    this->_send_offset += len;
+    this->_write_vio.ndone += len;
+  }
+  this->_records_stream_frame(level, *static_cast<QUICStreamFrame *>(frame));
+
+  this->_signal_write_event();
+  this->_state.update_with_sending_frame(*frame);
+
+  return frame;
+}
+
+QUICConnectionErrorUPtr
+QUICSendStream::recv(const QUICStopSendingFrame &frame)
+{
+  this->_state.update_with_receiving_frame(frame);
+  this->reset(QUICStreamErrorUPtr(new QUICStreamError(this, QUIC_APP_ERROR_CODE_STOPPING)));
+  // We received and processed STOP_SENDING frame, so return NO_ERROR here
+  return nullptr;
+}
+
+QUICConnectionErrorUPtr
+QUICSendStream::recv(const QUICMaxStreamDataFrame &frame)
+{
+  this->_remote_flow_controller.forward_limit(frame.maximum_stream_data());
+  QUICStreamFCDebug("[REMOTE] %" PRIu64 "/%" PRIu64, this->_remote_flow_controller.current_offset(),
+                    this->_remote_flow_controller.current_limit());
+
+  int64_t len = this->_process_write_vio();
+  if (len > 0) {
+    this->_signal_write_event();
+  }
+
+  return nullptr;
+}
+
+VIO *
+QUICSendStream::do_io_read(Continuation *c, int64_t nbytes, MIOBuffer *buf)
+{
+  QUICStreamDebug("Warning wants to read from send only stream ignore");
+  // FIXME: should not assert here
+  ink_assert(!"read from send only stream");
+  return nullptr;
```

Comment:
Not sure whether should assert here.

Reply:
Calling `QUICSendStream::do_io_read()` doesn't make sense, but VConnection assumes bidirectional. Make `do_io_read()` deleted function (`= delete`) could be another option.

Reply:
Yep, Let's keep this assertion since `do_io_read` is `VConnection`'s api. And It can not be mark as `delete` for an api which doesn't have the delete definition.

My only concern is the user might use the un-bidirectional stream as a `VConnection` . And it does not realize this VC is a un-bidirectional vc. In this scenario we should not assert here.

Reply:
I got point. At that case this should be just ignored. But probably, those Unidirectional Stream is used via QUICStreamIO APIs, so this assert doesn't make problem.